modules/ROOT/pages/mtf-metadata-testing.adoc

```diff
@@ -0,0 +1,363 @@
+        <munit-tools:assert-equals actual="#[payload.fields[0].model.'type']" expected="#['String']"/>
+        <munit-tools:assert-equals actual="#[payload.fields[1].key.name]" expected="#['name']"/>
+        <munit-tools:assert-equals actual="#[payload.fields[1].model.'type']" expected="#['String']"/>
+    </mtf:validation>
+</mtf:tooling-test>
+----
+
+=== Attributes Metadata
+
+The `attributes metadata` scope obtains the metadata type for the `attributes` metadata of the module. The `payload` contains a JSON representation of the metadata type:
+
+[source,xml,linenums]
+----
+<mtf:tooling-test name="attributesMetadataTest">
+    <mtf:get-attributes-metadata>
+        <example:my-operation config-ref="config" />
+    </mtf:get-attributes-metadata>
+    <mtf:validation>
+        <munit-tools:assert-equals actual="#[payload.format]" expected="#['java']"/>
+        <munit-tools:assert-equals actual="#[payload.fields[0].key.name]" expected="#['age']"/>
+        <munit-tools:assert-equals actual="#[payload.fields[0].model.'type']" expected="#['Number']"/>
+    </mtf:validation>
+</mtf:tooling-test>
+----
+
+The `mtf:get-values` element retrieves the possible values for the specified parameter of the module. The `payload` contains a JSON representation of the values.
+
+[source,xml,linenums]
+----
+<mtf:tooling-test name="attributesMetadataTest">
+    <mtf:get-values parameter="aParam">
+        <example:my-operation config-ref="config"/>
+    </mtf:get-values>
+    <mtf:validation>
+        <munit-tools:assert-equals actual="#[payload.WRITER.displayName]" expected="WRITER"/>
+        <munit-tools:assert-equals actual="#[payload.READER.displayName]" expected="READER"/>
+        <munit-tools:assert-equals actual="#[payload.ADMIN.displayName]" expected="ADMIN"/>
+    </mtf:validation>
+</mtf:tooling-test>
+----
+
+== Validation
+
+A common use case when testing metadata is validating failures that might occur during metadata calculation. To validate these failures, use the `expectFailureCode` and `expectFailureMessage` parameters:
+
+[source,xml,linenums]
+----
+<mtf:tooling-test name="expectFailureTest"
+        expectFailureCode="INVALID_METADATA_KEY"
+        expectFailureMessage="No metadata available for an empty name">
+    <mtf:get-output-metadata>
+        <example:my-operation config-ref="config" id=""/>
+    </mtf:get-output-metadata>
+</mtf:tooling-test>
+----
+
+=== Assert Type Processor
+
+Use the assert type processor to assert the metadata type structure returned by metadata scopes such as `get-input-metadata`, `get-output-metadata`, and `get-attributes-metadata`.
+
+The assert type processor compares the structure of both metadata types and ignores fields such as `format` or `annotations`. To perform assertions over these fields, you can use the `payload` result of the metadata scopes operations as shown in the previous examples.
+
+If the assertion fails, the processor throws a `java.lang.AssertionError` error.
+
+==== From Class
```

Comment:
This is getting deep, hierarchically. Maybe OK?

Reply:
Maybe a DL for these H4s?

Reply:
I just looked at it in context and you're right - there's enough stuff in each H4 to justify a section.